### DISCUSSION TITLE
Specify Euler angles for the three-grain model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#17196] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,

Recently, I modified the initial model of Tricrystal2CircleGrains based on [poly_grain_growth_2D_eldrforce.i](https://github.com/idaholab/moose/blob/next/modules/combined/examples/phase_field-mechanics/poly_grain_growth_2D_eldrforce.i). And the input file is 
[tri_grain_growth_2D_eldrforce.txt](https://github.com/idaholab/moose/files/6073221/tri_grain_growth_2D_eldrforce.txt)
I want to specify a fixed Euler angle for each grain, but the input file I write now uses a tex file to give the Euler angle different match. Is there any way to specify the Euler angle for each grain?

This is the specific content of the tex file，
```
Texture File
File generated from MATLAB
B 3
   30.00   0.00   0.00   1.00
   45.00   0.00   0.00   1.00
   90.00   0.00   0.00   1.00
```

The three-grain model and Euler angle are as follows
![image](https://user-images.githubusercontent.com/56183775/109743355-843ce580-7c0b-11eb-802a-f844d31cf5e5.png)

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thank you
Wei Peng

**Comment by ngrilli** [comment, marked as answer]
Dear @PengWei97 

Some of the examples in moose use the GrainPropertyReadFile object
to read the Euler angles element by element,
so each line of the Euler angle file corresponds to the angles in one element.

I have developed some simple modifications to assign the Euler angles
to different grains instead. This is done by considering different grains as different blocks.

You can see an example here with four grains:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/polycrystal

You can use this strategy to generate polycrystals with neper
and assign corresponding Euler angles grain by grain, as done here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/neper

Trust this helps

Best Regards,
Nicolò

**Reply by PengWei97** [reply]
Dear @ngrilli 

Thank you very much for your enthusiastic answers and the learning materials you provided. I have found the answer to this question. The problem lies in the *.tex file. Because the first Euler angle of the *.tex file I wrote is on the fourth line, this is a line that moose ignores, resulting in actually only two Euler angles specified, and the third Euler angle is assigned randomly. That's the problem.

Best Regards,
Wei Peng

**Reply by latmarat** [reply]
Hi Nicolò: 

Could you please comment how MOOSE orders the elements? That is, when you write the text file with the euler angles, how do you make sure it corresponds to the right element in the mesh? Does MOOSE sort elements such that element label increases first along x, then along y, then along z?

Thanks, 
Marat

**Reply by GiudGiud** [reply]
Hello

- if you are generating your mesh in MOOSE using the mesh generators, then there is some sort of ordering based on the direction, depending on the mesh generators you are using
- if you are generating your mesh outside of MOOSE, MOOSE does not re-order the elements (unless it is running in parallel. This should be disabled by passing allow_mesh_renumbering = false if you plan to set Euler angles by element). You should generate your Euler angles at the same time as you are generating your mesh to make sure the numbering match. 

Guillaume
